File: services/clinical-service/src/domain/entities/Patient.ts
import { v4 as uuidv4 } from 'uuid';

export enum Gender {
    MALE = 'MALE',
    FEMALE = 'FEMALE',
    OTHER = 'OTHER',
    UNKNOWN = 'UNKNOWN',
}

export interface ContactInfo {
    phone: string;
    email: string;
    address: string;
}

export interface PatientProps {
    id?: string;
    mrn: string;
    firstName: string;
    lastName: string;
    dateOfBirth: Date;
    gender: Gender;
    contactInfo: ContactInfo;
    createdAt?: Date;
    updatedAt?: Date;
}

export class Patient {
    readonly id: string;
    readonly mrn: string;
    readonly firstName: string;
    readonly lastName: string;
    readonly dateOfBirth: Date;
    readonly gender: Gender;
    readonly contactInfo: ContactInfo;
    readonly createdAt: Date;
    readonly updatedAt: Date;

    private constructor(props: Required<PatientProps>) {
        this.id = props.id;
        this.mrn = props.mrn;
        this.firstName = props.firstName;
        this.lastName = props.lastName;
        this.dateOfBirth = props.dateOfBirth;
        this.gender = props.gender;
        this.contactInfo = props.contactInfo;
        this.createdAt = props.createdAt;
        this.updatedAt = props.updatedAt;
    }

    static create(props: PatientProps): Patient {
        return new Patient({
            id: props.id ?? uuidv4(),
            mrn: props.mrn,
            firstName: props.firstName,
            lastName: props.lastName,
            dateOfBirth: props.dateOfBirth,
            gender: props.gender,
            contactInfo: props.contactInfo,
            createdAt: props.createdAt ?? new Date(),
            updatedAt: props.updatedAt ?? new Date(),
        });
    }

    getFullName(): string {
        return `${this.firstName} ${this.lastName}`;
    }
}

File: services/clinical-service/src/main/java/com/medtrust/clinical/domain/model/Patient.java
package com.medtrust.clinical.domain.model;

import java.time.Instant;
import java.util.UUID;

public class Patient {

    private final String id;
    private final String mrn;
    private final String firstName;
    private final String lastName;
    private final Instant dateOfBirth;
    private final Gender gender;
    private final ContactInfo contactInfo;
    private final Instant createdAt;
    private final Instant updatedAt;

    private Patient(String id, String mrn, String firstName, String lastName,
                    Instant dateOfBirth, Gender gender, ContactInfo contactInfo,
                    Instant createdAt, Instant updatedAt) {
        this.id = id;
        this.mrn = mrn;
        this.firstName = firstName;
        this.lastName = lastName;
        this.dateOfBirth = dateOfBirth;
        this.gender = gender;
        this.contactInfo = contactInfo;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }

    public static Patient create(String mrn, String firstName, String lastName,
                                  Instant dateOfBirth, Gender gender, ContactInfo contactInfo) {
        return new Patient(
                UUID.randomUUID().toString(),
                mrn,
                firstName,
                lastName,
                dateOfBirth,
                gender,
                contactInfo,
                Instant.now(),
                Instant.now()
        );
    }

    public static Patient reconstitute(String id, String mrn, String firstName, String lastName,
                                        Instant dateOfBirth, Gender gender, ContactInfo contactInfo,
                                        Instant createdAt, Instant updatedAt) {
        return new Patient(id, mrn, firstName, lastName, dateOfBirth, gender,
                contactInfo, createdAt, updatedAt);
    }

    public String getId() {
        return id;
    }

    public String getMrn() {
        return mrn;
    }

    public String getFirstName() {
        return firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public Instant getDateOfBirth() {
        return dateOfBirth;
    }

    public Gender getGender() {
        return gender;
    }

    public ContactInfo getContactInfo() {
        return contactInfo;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }

    public String getFullName() {
        return firstName + " " + lastName;
    }
}
